network/p2p/cache/node_blocklist_wrapper_test.go

```diff
@@ -0,0 +1,169 @@
+	"github.com/onflow/flow-go/network/p2p/cache"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+type NodeBlocklistWrapperTestSuite struct {
+	suite.Suite
+	DB       *badger.DB
+	provider *mocks.IdentityProvider
+
+	wrapper *cache.NodeBlocklistWrapper
+}
+
+func (s *NodeBlocklistWrapperTestSuite) SetupTest() {
+	s.DB, _ = unittest.TempBadgerDB(s.T())
+	s.provider = new(mocks.IdentityProvider)
+
+	var err error
+	s.wrapper, err = cache.NewNodeBlocklistWrapper(s.provider, s.DB)
+	require.NoError(s.T(), err)
+}
+
+func TestNodeBlocklistWrapperTestSuite(t *testing.T) {
+	suite.Run(t, new(NodeBlocklistWrapperTestSuite))
+}
+
+// TestHonestNode verifies:
+// For nodes _not_ on the blocklist, the `cache.NodeBlocklistWrapper` should forward
+// the identities from the wrapped `IdentityProvider` without modification.
+func (s *NodeBlocklistWrapperTestSuite) TestHonestNode() {
+	s.Run("ByNodeID", func() {
+		identity := unittest.IdentityFixture()
+		s.provider.On("ByNodeID", identity.NodeID).Return(identity, true)
+
+		i, found := s.wrapper.ByNodeID(identity.NodeID)
+		require.True(s.T(), found)
+		require.Equal(s.T(), i, identity)
+	})
+	s.Run("ByPeerID", func() {
+		identity := unittest.IdentityFixture()
+		peerID := (peer.ID)("some_peer_ID")
+		s.provider.On("ByPeerID", peerID).Return(identity, true)
+
+		i, found := s.wrapper.ByPeerID(peerID)
+		require.True(s.T(), found)
+		require.Equal(s.T(), i, identity)
+	})
+	s.Run("Identities", func() {
+		identities := unittest.IdentityListFixture(11)
+		filter := filter.In(identities[3:4])
+		expectedFilteredIdentities := identities.Filter(filter)
+		s.provider.On("Identities", mock.Anything).Return(
+			func(filter flow.IdentityFilter) flow.IdentityList {
+				return identities.Filter(filter)
+			},
+			nil,
+		)
+		require.Equal(s.T(), expectedFilteredIdentities, s.wrapper.Identities(filter))
+	})
+}
+
+// TestBlacklistedNode tests proper handling of identities _on_ the blocklist:
+//   - For any identity `i` with `i.NodeID ∈ blocklist`, the returned identity
+//     should have `i.Ejected` set to `true` (irrespective of the `Ejected`
+//     flag's initial returned by the wrapped `IdentityProvider`).
+//   - The wrapper should _copy_ the identity and _not_ write into the wrapped
+//     IdentityProvider's memory.
+//   - For `IdentityProvider.ByNodeID` and `IdentityProvider.ByPeerID`:
+//     whether or not the wrapper modifies the `Ejected` flag should depend only
+//     in the NodeID of the returned identity, irrespective of the second return
+//     value (boolean).
+//     While returning (non-nil identity, false) is not a defined return value,
+//     we expect the wrapper to nevertheless handle this case to increase its
+//     generality.
+func (s *NodeBlocklistWrapperTestSuite) TestBlacklistedNode() {
+	blocklist := unittest.IdentityListFixture(11)
+	err := s.wrapper.Update(blocklist.NodeIDs())
+	require.NoError(s.T(), err)
+
+	index := atomic.NewInt32(0)
+	for _, b := range []bool{true, false} {
```

Comment:
Based on the documentation of the `IdentityProvider` interface, the boolean return value of the `ByPeerID` and `ByNodeID` is _"true if and only if Identity has been found, i.e. `Identity` is not nil."_. Accordingly, there will be no happy path case where we have the combination of a non-nil identity and `false` as the returned value. Along the same line, it seems we don't need to iterate over `[]bool{true, false}` rather, all _boolean_ returned values for these functions in this test must be set to `true` (and `false` for the subsequent `TestUnknownNode`).

Reply:
Completely agree that from the perspective of `IdentityProvider` the return value `(<non-nil identity>, false)` is not specified. 
However, I beliefe there i